### PR TITLE
Classes for receipts transfer (not wired up)

### PR DESCRIPTION
**Pull request by aion-kelvin**
## Notice

It is not allowed to submit your PR to the master branch directly, please submit your PR to the master-pre-merge branch.

## Description

Please include a brief summary of the change that this pull request proposes. Include any relevant motivation and context. List any dependencies required for this change.

- This is for the receipts transfer component of Fast Sync
- The PR contains:
  - Request/response classes for receipt transfer requests
  - Request handler and response handler for those classes
  - Classes for aiding testing/verification of the receipt transfer
  - Changes to AionHub and TaskImportBlocks to hook in the new handlers  and verification helpers
  - Integ tests -- they do not run as part of CI, but I ran them locally 
- In this PR, the enable flag for requesting receipt transfers is hardcoded to be disabled (see AionHub.java:183), but the handler for requests/responses is enabled.
- The code in this PR is not what the final form of what receipts transfer should look like, some refactoring will needed for that.  The following things will need to change:
  - Remove hardcoded disabling of receipt transfers (AionHub.java:183)
  - Right now, ReceiptsRetrievalVerifier sends the request for receipts transfers.  This class is meant to be an intermediate step used for testing, so we will eventually need to move that logic into TaskImportblocks or something else.  Don't want to do this right now because this verifier will be handy for integrating this code with the rest of fast sync and also, not sure if TaskImportBlocks is where we want this to hook in
  - Disabling of normal receipts computation -- currently, turning on receipt requests does not turn off normal computation of receipts

Fixes Issue # .

## Type of change

Insert **x** into the following checkboxes to confirm (eg. [x]):
- [ ] Bug fix.
- [ ] New feature.
- [ ] Enhancement.
- [ ] Unit test.
- [ ] Breaking change (a fix or feature that causes existing functionality to not work as expected).
- [ ] Requires documentation update.

## Testing

Please describe the tests you used to validate this pull request. Provide any relevant details for test configurations as well as any instructions to reproduce these results.

- The two new integration tests pass
- This code was written back in Oct/2018, during which time, the output from ReceiptsRetrievalVerifier was used to confirm that:
  - after running kernel over night, all requested receipts were fulfilled
  - requested receipts were added to an alternate database, then checked against the real tranactions db to ensure that they are the same 

## Verification

Insert **x** into the following checkboxes to confirm (eg. [x]):
- [ ] I have self-reviewed my own code and conformed to the style guidelines of this project.
- [ ] New and existing tests pass locally with my changes.
- [ ] I have added tests for my fix or feature.
- [ ] I have made appropriate changes to the corresponding documentation.
- [ ] My code generates no new warnings.
- [ ] Any dependent changes have been made.

### CODE REVIEW

File: modAionImpl/src/org/aion/zero/impl/sync/Act.java [hunk 1]
```diff
@@ -22,4 +22,8 @@
     public static final byte REQUEST_TRIE_DATA = 8;
 
     public static final byte RESPONSE_TRIE_DATA = 9;
+
+    public static final byte REQ_TX_RECEIPT_HEADERS = 10; // TODO ??
+
+    public static final byte RES_TX_RECEIPT_HEADERS = 11; // TODO ??
```
Comment (AlexandraRoatis) [inline modAionImpl/src/org/aion/zero/impl/sync/Act.java]: please rename the constants to `REQUEST_RECEIPTS` and `RESPONSE_RECEIPTS` and remove the comments

File: modAionImpl/src/org/aion/zero/impl/AionHub.java [hunk 2]
```diff
@@ -505,6 +525,10 @@ public void close() {
             genLOG.info("shutdown DB... Done!");
         }
 
+        if(instrumentedResTxReceiptHandler != null) {
```
Comment (AlexandraRoatis) [inline modAionImpl/src/org/aion/zero/impl/AionHub.java]: formatting

File: modAionImpl/src/org/aion/zero/impl/sync/ReceiptsRetrievalVerifier.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Verify receipts requests process by:
+ *  1) comparing the list of tx receipts we've requested against the list of tx receipts we've received.
+ *  2) checking that the received receipt matches that of the receipt we would have made if we had computed it locally
```
Comment (AlexandraRoatis) [inline modAionImpl/src/org/aion/zero/impl/sync/ReceiptsRetrievalVerifier.java]: Formatting will make this harder to read, but you can use html tags to create an ordered list.

File: modAionImpl/src/org/aion/zero/impl/sync/msg/ReqTxReceipts.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+import java.util.List;
+
+import com.google.common.base.Preconditions;
+import org.aion.p2p.Ctrl;
+import org.aion.p2p.Msg;
+import org.aion.p2p.Ver;
+import org.aion.zero.impl.sync.Act;
+
+/** Request for transaction receipts */
+public class ReqTxReceipts extends Msg {
```
Comment (AlexandraRoatis) [inline modAionImpl/src/org/aion/zero/impl/sync/msg/ReqTxReceipts.java]: please rename this to `RequestReceipts`; same for the handler and tests classes

File: modAionImpl/src/org/aion/zero/impl/sync/msg/ResTxReceipts.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+import org.aion.zero.impl.sync.Act;
+import org.aion.zero.impl.types.AionTxInfo;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+
+/** Response for transaction receipts request */
+public class ResTxReceipts extends Msg {
```
Comment (AlexandraRoatis) [inline modAionImpl/src/org/aion/zero/impl/sync/msg/ResTxReceipts.java]: please rename this to `ResponseReceipts`; same for the handler and tests classes

File: modP2pImpl/src/module-info.java [hunk 6]
```diff
@@ -7,6 +7,7 @@
     requires jsr305;
     requires commons.collections4;
 
+
```
Comment (AlexandraRoatis) [inline modP2pImpl/src/module-info.java]: unnecessary change

File: modP2pImpl/src/org/aion/p2p/impl1/tasks/TaskReceive.java [hunk 7]
```diff
@@ -41,6 +41,7 @@ public void run() {
                     try {
                         hlr.receive(mi.getNodeId(), mi.getDisplayId(), mi.getMsg());
                     } catch (Exception e) {
+                        e.printStackTrace();
```
Comment (AlexandraRoatis) [inline modP2pImpl/src/org/aion/p2p/impl1/tasks/TaskReceive.java]: Why print the stack trace here?

File: modP2pImpl/src/org/aion/p2p/impl1/tasks/TaskWrite.java [hunk 8]
```diff
@@ -1,14 +1,15 @@
...
 
 import java.io.IOException;
 import java.nio.ByteBuffer;
 import java.nio.channels.ClosedChannelException;
 import java.nio.channels.SocketChannel;
-import org.aion.p2p.Header;
-import org.aion.p2p.IP2pMgr;
-import org.aion.p2p.Msg;
+
+import static org.aion.p2p.impl1.P2pMgr.p2pLOG;
```
Comment (AlexandraRoatis) [inline modP2pImpl/src/org/aion/p2p/impl1/tasks/TaskWrite.java]: changes not necessary

File: modP2pImpl/src/org/aion/p2p/impl1/tasks/TaskReceive.java (hunk 7 above)
Comment (aion-kelvin) [reply modP2pImpl/src/org/aion/p2p/impl1/tasks/TaskReceive.java]: oops, leftover from when i was prototyping

**Comment by AionJayT** [comment]
change milestone to 0.4

**Comment by arajasek** [comment]
Thanks for working on this! Two minor changes suggested privately, regarding `ReceiptsRetrievalVerifier.displayOutstandingRequests()` and the `AionHub` "changes". 

Everything else looks good! :)

**Comment by aion-kelvin** [comment]
will cut a new PR containing this functionality when block transfer code is in master-pre-merge (not clear where to integrate this right now / rebasing too much for throwaway code)
